File: sokoban-cpp/Metrics.cpp
//
// Created by atopi on 02.12.2018.
//

#include "Metrics.h"
#include <queue>
#include <cmath>
#include <iostream>

Metrics::Metrics(int width, int height, int *targets, int targets_count, int *clearedBoard)
{
    Metrics::width = width;
    Metrics::height = height;
    Metrics::size = width*height;
    Metrics::targets_count = targets_count;
    Metrics::target_codes = new int[targets_count];

    Metrics::distance_to_goals = new int*[size];
    for(int i = 0; i < size; i++)
    {
        Metrics::distance_to_goals[i] = new int[size];
        for(int j = 0; j < size; j++)
            Metrics::distance_to_goals[i][j] = 32000;
    }

    /*std::queue<int> target_queue = std::queue<int>();

    for(int i = 0; i < targets_count; i++)
    {
        distance_to_goals[i][targets[i]] = 0;
        target_codes[i] = targets[i];
        target_queue.push(targets[i]);
        while(!target_queue.empty())
        {
            int pos = target_queue.front();
            target_queue.pop();
            for(int d : {-1, 1, -width, width})
            {
                int box_pos = pos + d;
                int player_pos = pos + 2*d;
                if(distance_to_goals[i][box_pos] == 32000 && clearedBoard[box_pos] != 4 && clearedBoard[player_pos] != 4)
                {
                    distance_to_goals[i][box_pos] = distance_to_goals[i][pos] + 1;
                    target_queue.push(box_pos);
                }
            }
        }
    }*/

    std::queue<int> mqueue = std::queue<int>();
    for(int i = 0; i < size; i++)
    {
        distance_to_goals[i][i] = 0;
        mqueue.push(i);

        while(!mqueue.empty())
        {
            int position = mqueue.front(); mqueue.pop();
            for(int d : {-1, 1, -width, width})
            {
                int player_pos = position + d;
                if(player_pos < 0 || player_pos >= size)
                    continue;
                if(distance_to_goals[i][player_pos] == 32000 && clearedBoard[player_pos] != 4)
                {
                    distance_to_goals[i][player_pos] = distance_to_goals[i][position] +1;
                    mqueue.push(player_pos);
                }
            }
        }
    }
}

Metrics::~Metrics()
{
    for(int i = 0; i < size; i++)
        delete distance_to_goals[i];
    delete distance_to_goals;
}

int Metrics::manhattan_distance(int a, int b, int width)
{
    return (int)(std::abs((a % width) - (a / width)) + std::abs((b % width) - (b / width)));
}

int Metrics::pythagoran_distance(int a, int b)
{
    return (int)(std::sqrt(((a % width) - (b % width))*((a % width) - (b % width)) + ((a / width) - (b / width))*((a / width) - (b / width))));
}

int Metrics::lookupTarget(int target)
{
    for(int i = 0; i < targets_count; i++)
        if(target_codes[i] == target)
            return i;

    return 0;
}

int Metrics::pull_distance(int target, int pos)
{
    return Metrics::distance_to_goals[target][pos];
}

File: sokoban-python/Metrics.py
import math
import numpy as np
import queue

# Metric codes
PYTHAGOREAN_METRIC = 1
MANHATTAN_METRIC = 2
PULL_METRIC = 3


class Metrics:

    def __init__(self, width, height, targets, clearedBoard, metric_code):
        self.width = width
        self.height = height
        self.size = width * height
        self.metric_code = metric_code
        self.distances = np.full((self.size, self.size), 32000, dtype=int)

        mqueue = queue.Queue()
        for i in range(0, self.size, 1):
            self.distances[i][i] = 0
            mqueue.put(i)
            while not mqueue.empty():
                position = mqueue.get()
                for d in [-1, 1, -self.width, self.width]:
                    player_pos = position + d
                    if not 0 <= player_pos < self.size:
                        continue
                    if self.distances[i][player_pos] == 32000 and not clearedBoard[player_pos] == 4:
                        self.distances[i][player_pos] = self.distances[i][position] + 1
                        mqueue.put(player_pos)


    @staticmethod
    def manhattan_distance(a, b, width):
        a_x = int(a % width)
        a_y = int(a / width)
        b_x = int(b % width)
        b_y = int(b / width)
        return Metrics.manhattan_distance_2d(a_x, a_y, b_x, b_y)

    @staticmethod
    def manhattan_distance_2d(a_x, a_y, b_x, b_y):
        return abs(a_x - b_x) + abs(a_y - b_y)

    def pythagorean_distance(self, a, b):
        a_x = int(a % self.width)
        a_y = int(a / self.width)
        b_x = int(b % self.width)
        b_y = int(b / self.width)
        return Metrics.pythagorean_distance_2d(a_x, a_y, b_x, b_y)

    @staticmethod
    def pythagorean_distance_2d(a_x, a_y, b_x, b_y):
        return math.sqrt((a_x - b_x)**2 + (a_y - b_y)**2)

    def pull_distance(self, to, pos):
        if not 0 <= to < self.size or not 0 <= pos < self.size:
            return -1
        else:
            return self.distances[to][pos]

    def pull_distance_2d(self, target_x, target_y, pos_x, pos_y):
        target = self.width * target_y + target_x
        pos = self.width * pos_x + pos_y
        return self.pull_distance(target, pos)

    def distance(self, target, pos):
        if self.metric_code == PYTHAGOREAN_METRIC:
            return self.pythagorean_distance(target, pos)
        if self.metric_code == MANHATTAN_METRIC:
            return self.manhattan_distance(target, pos, self.width)
        if self.metric_code == PULL_METRIC:
            return self.pull_distance(target, pos)

    def distance_2d(self, target_x, target_y, pos_x, pos_y):
        if self.metric_code == PYTHAGOREAN_METRIC:
            return self.pythagorean_distance_2d(target_x, target_y, pos_x, pos_y)
        if self.metric_code == MANHATTAN_METRIC:
            return self.manhattan_distance_2d(target_x, target_y, pos_x, pos_y)
        if self.metric_code == PULL_METRIC:
            return self.pull_distance_2d(target_x, target_y, pos_x, pos_y)
